lib/Raven/ErrorHandler.php

```diff
@@ -11,203 +9,295 @@
+
+        if (!$shouldReportError && !$shouldCaptureError) {
+            return false;
         }
 
-        if ($this->call_existing_error_handler) {
-            if (null !== $this->old_error_handler) {
-                return call_user_func(
-                    $this->old_error_handler,
-                    $type,
-                    $message,
-                    $file,
-                    $line,
-                    $context
-                );
-            } else {
-                return false;
-            }
+        $errorAsException = new \ErrorException(self::$errorLevelsDescriptions[$level] . ': ' . $message, 0, $level, $file, $line);
+
+        $backtrace = $errorAsException->getTrace();
+        $backtrace = $this->cleanBacktraceFromErrorHandlerFrames($backtrace, $file, $line);
+
+        $this->exceptionReflection->setValue($errorAsException, $backtrace);
+
+        try {
+            $this->handleException($errorAsException);
+        } catch (\Exception $exception) {
+            // Do nothing as this error handler should be as trasparent as possible
+        }
+
+        if (null !== $this->previousErrorHandler) {
+            return call_user_func($this->previousErrorHandler, $level, $message, $file, $line);
         }
 
-        return true;
+        return $shouldReportError;
     }
 
-    public function handleFatalError()
+    /**
+     * Handles fatal errors by capturing them through the Raven client. This
+     * method is used as callback of a shutdown function.
+     *
+     * @param array|null $error The error details as returned by error_get_last()
+     *
+     * @internal
+     */
+    public function handleFatalError(array $error = null)
     {
-        unset($this->reservedMemory);
-
-        if (null === $error = error_get_last()) {
+        // If there is not enough memory that can be used to handle the error
+        // do nothing
+        if (null === self::$reservedMemory) {
             return;
         }
 
-        if ($this->shouldCaptureFatalError($error['type'])) {
-            $e = new \ErrorException(
-                @$error['message'],
-                0,
-                @$error['type'],
-                @$error['file'],
-                @$error['line']
-            );
-            $this->handleException($e, true);
+        self::$reservedMemory = null;
+        $exception = null;
+
+        if (null === $error) {
+            $error = error_get_last();
         }
-    }
 
-    public function shouldCaptureFatalError($type)
-    {
-        // Do not capture E_ERROR since those can be caught by userland since PHP 7.0
-        // E_ERROR should already be handled by the exception handler
-        // This prevents duplicated exceptions in PHP 7.0+
-        if (PHP_VERSION_ID >= 70000 && E_ERROR === $type) {
-            return false;
+        if (!empty($error) && $error['type'] & (E_ERROR | E_PARSE | E_CORE_ERROR | E_CORE_WARNING | E_COMPILE_ERROR | E_COMPILE_WARNING)) {
```

Comment:
Maybe you should wrap those two lines with a private method to give it a name?

Reply:
Extracting the code that just checks whether the error is a fatal error does not makes much sense imho, and there would also be the cost of calling one more function